ConcurrentModificationException while removing more then 1 potion effect

Glowstone commit: https://github.com/GlowstoneMC/Glowstone/commit/b3b6502f64772bdde3d397c29e5d4d0386c06836
If try remove all player effects and he already have more the 1 potion effect active `ConcurrentModificationException` throw.
Can be reproduced with command `/effect name clear` or code 
```java
for (PotionEffect potionEffect : player.getActivePotionEffects()) {
     player.removePotionEffect(potionEffect.getType());
}
```
Stacktrace:
```
[SEVERE] Exception while executing command: /effect SHADOWDAN_ clear
org.bukkit.command.CommandException: Unhandled exception executing 'effect SHADOWDAN_ clear' in net.glowstone.command.minecraft.EffectCommand(effect)
        at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:161)
        at net.glowstone.GlowServer.dispatchCommand(GlowServer.java:1903)
        at net.glowstone.entity.GlowPlayer.lambda$chat$15(GlowPlayer.java:2286)
        at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at java.util.concurrent.FutureTask.run(Unknown Source)
        at net.glowstone.scheduler.GlowTask.run(GlowTask.java:144)
        at net.glowstone.scheduler.GlowScheduler.pulse(GlowScheduler.java:196)
        at net.glowstone.scheduler.GlowScheduler.lambda$start$0(GlowScheduler.java:122)
        at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at java.util.concurrent.FutureTask.runAndReset(Unknown Source)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(Unknown Source)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: java.util.ConcurrentModificationException
        at java.util.HashMap$HashIterator.nextNode(Unknown Source)
        at java.util.HashMap$ValueIterator.next(Unknown Source)
        at java.util.Collections$UnmodifiableCollection$1.next(Unknown Source)
        at net.glowstone.command.minecraft.EffectCommand.execute(EffectCommand.java:67)
        at net.glowstone.command.minecraft.GlowVanillaCommand.execute(GlowVanillaCommand.java:155)
        at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:151)
        ... 14 more
```

--- a/src/main/java/net/glowstone/GlowWorld.java
+++ b/src/main/java/net/glowstone/GlowWorld.java
@@ -2014,11 +2014,11 @@ public class GlowWorld implements World {
             getRawPlayers().forEach(GlowPlayer::sendTime);
         } else if (rule.equals(GameRules.REDUCED_DEBUG_INFO)) {
             // inform clients about the debug info change
-            EntityStatusMessage message = new EntityStatusMessage(0,
-                gameRuleMap.getBoolean(GameRules.REDUCED_DEBUG_INFO)
-                    ? EntityStatusMessage.ENABLE_REDUCED_DEBUG_INFO
-                    : EntityStatusMessage.DISABLE_REDUCED_DEBUG_INFO);
             for (GlowPlayer player : getRawPlayers()) {
+                EntityStatusMessage message = new EntityStatusMessage(player.getEntityId(),
+                        gameRuleMap.getBoolean(GameRules.REDUCED_DEBUG_INFO)
+                                ? EntityStatusMessage.ENABLE_REDUCED_DEBUG_INFO
+                                : EntityStatusMessage.DISABLE_REDUCED_DEBUG_INFO);
                 player.getSession().send(message);
             }
         }

--- a/src/main/java/net/glowstone/entity/GlowEntity.java
+++ b/src/main/java/net/glowstone/entity/GlowEntity.java
@@ -1166,6 +1166,17 @@ public abstract class GlowEntity implements Entity {
         }
     }
 
+    public void playEffectKnownAndSelf(EntityEffect type) {
+        if (type.getApplicable().isInstance(this)) {
+            EntityStatusMessage message = new EntityStatusMessage(entityId, type);
+            if (this instanceof GlowPlayer) {
+                ((GlowPlayer) this).getSession().send(message);
+            }
+            world.getRawPlayers().stream().filter(player -> player.canSeeEntity(this))
+                    .forEach(player -> player.getSession().send(message));
+        }
+    }
+
     @Override
     public EntityType getType() {
         return EntityType.UNKNOWN;

--- a/src/main/java/net/glowstone/entity/GlowEntity.java
+++ b/src/main/java/net/glowstone/entity/GlowEntity.java
@@ -850,13 +850,7 @@ public abstract class GlowEntity implements Entity {
             // In case the current session belongs to this player passenger
             // We need to send the self_id
             List<Integer> passengerIds = new ArrayList<>();
-            getPassengers().forEach(e -> {
-                if (session.getPlayer().equals(e)) {
-                    passengerIds.add(GlowPlayer.SELF_ID);
-                } else {
-                    passengerIds.add(e.getEntityId());
-                }
-            });
+            getPassengers().forEach(e -> passengerIds.add(e.getEntityId()));
             result.add(new SetPassengerMessage(getEntityId(), passengerIds.stream()
                 .mapToInt(Integer::intValue).toArray()));
             passengerChanged = false;
@@ -1165,9 +1159,11 @@ public abstract class GlowEntity implements Entity {
 
     @Override
     public void playEffect(EntityEffect type) {
-        EntityStatusMessage message = new EntityStatusMessage(entityId, type);
-        world.getRawPlayers().stream().filter(player -> player.canSeeEntity(this))
-            .forEach(player -> player.getSession().send(message));
+        if (type.getApplicable().isInstance(this)) {
+            EntityStatusMessage message = new EntityStatusMessage(entityId, type);
+            world.getRawPlayers().stream().filter(player -> player.canSeeEntity(this))
+                    .forEach(player -> player.getSession().send(message));
+        }
     }
 
     @Override

--- a/src/main/java/net/glowstone/entity/GlowLivingEntity.java
+++ b/src/main/java/net/glowstone/entity/GlowLivingEntity.java
@@ -65,6 +65,7 @@ import org.bukkit.block.BlockFace;
 import org.bukkit.entity.Entity;
 import org.bukkit.entity.EntityType;
 import org.bukkit.entity.Fireball;
+import org.bukkit.entity.HumanEntity;
 import org.bukkit.entity.LivingEntity;
 import org.bukkit.entity.Monster;
 import org.bukkit.entity.Player;
@@ -74,6 +75,7 @@ import org.bukkit.event.entity.EntityDamageByEntityEvent;
 import org.bukkit.event.entity.EntityDamageEvent;
 import org.bukkit.event.entity.EntityDamageEvent.DamageCause;
 import org.bukkit.event.entity.EntityDeathEvent;
+import org.bukkit.event.entity.EntityResurrectEvent;
 import org.bukkit.event.entity.EntityToggleGlideEvent;
 import org.bukkit.event.entity.PlayerDeathEvent;
 import org.bukkit.event.entity.PlayerLeashEntityEvent;
@@ -832,29 +834,19 @@ public abstract class GlowLivingEntity extends GlowEntity implements LivingEntit
             return;
         }
 
+        if (this.tryUseTotem()) {
+            return;
+        }
+
         // Killed
         active = false;
         Sound deathSound = getDeathSound();
         if (deathSound != null && !isSilent()) {
             world.playSound(location, deathSound, getSoundVolume(), getSoundPitch());
         }
-        playEffect(EntityEffect.DEATH);
+        playEffectKnownAndSelf(EntityEffect.DEATH);
         if (this instanceof GlowPlayer) {
             GlowPlayer player = (GlowPlayer) this;
-            ItemStack mainHand = player.getInventory().getItemInMainHand();
-            ItemStack offHand = player.getInventory().getItemInOffHand();
-            if (!InventoryUtil.isEmpty(mainHand) && mainHand.getType() == Material.TOTEM) {
-                player.getInventory().setItemInMainHand(InventoryUtil.createEmptyStack());
-                player.setHealth(1.0);
-                active = true;
-                return;
-            }
-            if (!InventoryUtil.isEmpty(offHand) && offHand.getType() == Material.TOTEM) {
-                player.getInventory().setItemInOffHand(InventoryUtil.createEmptyStack());
-                player.setHealth(1.0);
-                active = true;
-                return;
-            }
             List<ItemStack> items = null;
             boolean dropInventory = !world.getGameRuleMap().getBoolean(GameRules.KEEP_INVENTORY);
             if (dropInventory) {
@@ -1001,7 +993,7 @@ public abstract class GlowLivingEntity extends GlowEntity implements LivingEntit
         }
 
         setHealth(health - amount);
-        playEffect(EntityEffect.HURT);
+        playEffectKnownAndSelf(EntityEffect.HURT);
 
         if (cause == DamageCause.ENTITY_ATTACK && source != null) {
             Vector distance = RayUtil
@@ -1174,6 +1166,12 @@ public abstract class GlowLivingEntity extends GlowEntity implements LivingEntit
         return Collections.unmodifiableCollection(potionEffects.values());
     }
 
+    public void clearActivePotionEffects() {
+        for (PotionEffect effect : this.getActivePotionEffects()) {
+            this.removePotionEffect(effect.getType());
+        }
+    }
+
     @Override
     public void setOnGround(boolean onGround) {
         float fallDistance = getFallDistance();
@@ -1330,5 +1328,45 @@ public abstract class GlowLivingEntity extends GlowEntity implements LivingEntit
         }
 
         return false;
+    }
+
+    /**
+     * Use "Totem of Undying" if equipped
+     * @return result of totem use
+     */
+    public boolean tryUseTotem() {
+        //TODO: Should return false if player die in void.
+        if (!(this instanceof HumanEntity)) {
+            return false;
+        }
+
+        HumanEntity human = (HumanEntity) this;
+        ItemStack mainHand = human.getInventory().getItemInMainHand();
+        ItemStack offHand = human.getInventory().getItemInOffHand();
+
+        boolean hasTotem = false;
+        if (!InventoryUtil.isEmpty(mainHand) && mainHand.getType() == Material.TOTEM) {
+            mainHand.setAmount(mainHand.getAmount() - 1);
+            human.getInventory().setItemInMainHand(InventoryUtil.createEmptyStack());
+            hasTotem = true;
+        } else if (!InventoryUtil.isEmpty(offHand) && offHand.getType() == Material.TOTEM) {
+            human.getInventory().setItemInOffHand(InventoryUtil.createEmptyStack());
+            hasTotem = true;
+        }
+
+        EntityResurrectEvent event = EventFactory.getInstance().callEvent(new EntityResurrectEvent(this));
+        event.setCancelled(!hasTotem);
+
+        if (event.isCancelled()) {
+            return false;
+        }
+
+        this.setHealth(1.0F);
+        this.clearActivePotionEffects();
+        this.addPotionEffect(new PotionEffect(PotionEffectType.REGENERATION, 900, 1));
+        this.addPotionEffect(new PotionEffect(PotionEffectType.ABSORPTION, 900, 1));
+        playEffectKnownAndSelf(EntityEffect.TOTEM_RESURRECT);
+
+        return true;
     }
 }

--- a/src/main/java/net/glowstone/entity/GlowLivingEntity.java
+++ b/src/main/java/net/glowstone/entity/GlowLivingEntity.java
@@ -8,13 +8,13 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
-import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Iterator;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ThreadLocalRandom;
 import java.util.stream.Collectors;
 import lombok.Getter;
@@ -132,7 +132,7 @@ public abstract class GlowLivingEntity extends GlowEntity implements LivingEntit
     /**
      * Potion effects on the entity.
      */
-    private final Map<PotionEffectType, PotionEffect> potionEffects = new HashMap<>();
+    private final Map<PotionEffectType, PotionEffect> potionEffects = new ConcurrentHashMap<>();
     /**
      * The LivingEntity's AttributeManager.
      */

--- a/src/main/java/net/glowstone/entity/GlowPlayer.java
+++ b/src/main/java/net/glowstone/entity/GlowPlayer.java
@@ -217,7 +217,6 @@ public class GlowPlayer extends GlowHumanEntity implements Player {
     /**
      * A static entity id to use when telling the client about itself.
      */
-    public static final int SELF_ID = 0;
     public static final int HOOK_MAX_DISTANCE = 32;
 
     private static final Achievement[] ACHIEVEMENT_VALUES = Achievement.values();
@@ -665,17 +664,21 @@ public class GlowPlayer extends GlowHumanEntity implements Player {
      * @param reader the source of the player's saved state
      */
     public void join(GlowSession session, PlayerReader reader) {
-        // send join game
-        // in future, handle hardcore, difficulty, and level type
         String type = world.getWorldType().getName().toLowerCase();
+
+        reader.readData(this);
+        reader.close();
+
         int gameMode = getGameMode().getValue();
         if (server.isHardcore()) {
             gameMode |= 0x8;
         }
-        session.send(new JoinGameMessage(SELF_ID, gameMode, world.getEnvironment().getId(), world
+
+        setGameModeDefaults();
+
+        session.send(new JoinGameMessage(getEntityId(), gameMode, world.getEnvironment().getId(), world
                 .getDifficulty().getValue(), session.getServer().getMaxPlayers(), type, world
                 .getGameRuleMap().getBoolean(GameRules.REDUCED_DEBUG_INFO)));
-        setGameModeDefaults();
 
         // send server brand and supported plugin channels
         Message pluginMessage = PluginMessage.fromString("MC|Brand", server.getName());
@@ -684,8 +687,6 @@ public class GlowPlayer extends GlowHumanEntity implements Player {
         }
         sendSupportedChannels();
         joinTime = System.currentTimeMillis();
-        reader.readData(this);
-        reader.close();
 
         // Add player to list of online players
         getServer().setPlayerOnline(this, true);
@@ -921,7 +922,7 @@ public class GlowPlayer extends GlowHumanEntity implements Player {
         // send changed metadata
         List<MetadataMap.Entry> changes = metadata.getChanges();
         if (!changes.isEmpty()) {
-            session.send(new EntityMetadataMessage(SELF_ID, changes));
+            session.send(new EntityMetadataMessage(getEntityId(), changes));
         }
 
         // Entity IDs are only unique per world, so we can't spawn or teleport between worlds while
@@ -971,7 +972,7 @@ public class GlowPlayer extends GlowHumanEntity implements Player {
         }
 
         if (passengerChanged) {
-            session.send(new SetPassengerMessage(SELF_ID, getPassengers().stream()
+            session.send(new SetPassengerMessage(getEntityId(), getPassengers().stream()
                     .mapToInt(Entity::getEntityId).toArray()));
         }
         getAttributeManager().sendMessages(session);
@@ -1354,7 +1355,7 @@ public class GlowPlayer extends GlowHumanEntity implements Player {
         if (!event.isCancelled()) {
             velocity = event.getVelocity();
             super.setVelocity(velocity);
-            session.send(new EntityVelocityMessage(SELF_ID, velocity));
+            session.send(new EntityVelocityMessage(getEntityId(), velocity));
         }
     }
 
@@ -2050,9 +2051,9 @@ public class GlowPlayer extends GlowHumanEntity implements Player {
         sleeping = true;
         setRawLocation(head.getLocation(), false);
 
-        getSession().send(new UseBedMessage(SELF_ID, head.getX(), head.getY(), head.getZ()));
         UseBedMessage msg = new UseBedMessage(getEntityId(), head.getX(), head.getY(), head.getZ());
-        world.getRawPlayers().stream().filter(p -> p != this && p.canSeeEntity(this))
+        getSession().send(msg);
+        world.getRawPlayers().stream().filter(p -> p.canSeeEntity(this))
                 .forEach(p -> p.getSession().send(msg));
     }
 
@@ -3633,7 +3634,7 @@ public class GlowPlayer extends GlowHumanEntity implements Player {
 
     @Override
     public void playAnimationToSelf(EntityAnimation animation) {
-        AnimateEntityMessage message = new AnimateEntityMessage(SELF_ID, animation.ordinal());
+        AnimateEntityMessage message = new AnimateEntityMessage(getEntityId(), animation.ordinal());
         getSession().send(message);
     }
 

--- a/src/main/java/net/glowstone/entity/passive/GlowFishingHook.java
+++ b/src/main/java/net/glowstone/entity/passive/GlowFishingHook.java
@@ -61,13 +61,11 @@ public class GlowFishingHook extends GlowProjectile implements FishHook {
                     .filter(player -> player.canSeeEntity(this))
                     .forEach(player -> player.getSession().sendAll(
                             respawnMessages.toArray(EMPTY_MESSAGE_ARRAY)));
-            // Each player believes her own entity ID is 0, so the update is sent separately to the
-            // shooter
             if (shooter instanceof GlowPlayer) {
                 GlowSession session = ((GlowPlayer) shooter).getSession();
                 session.send(destroyOldCopy);
                 session.sendAll(
-                        createSpawnMessage(GlowPlayer.SELF_ID).toArray(EMPTY_MESSAGE_ARRAY));
+                        createSpawnMessage(getEntityId()).toArray(EMPTY_MESSAGE_ARRAY));
             }
         }
     }

